scylla-cql/src/frame/value.rs

```diff
@@ -230,6 +215,78 @@ impl<'a> Iterator for SerializedValuesIterator<'a> {
     }
 }
 
+/// Represents List of ValueList for Batch statement
+///
+/// This trait is not implemented directly, but rather implemented through `BatchValuesGatWorkaround`
+/// (until GATs are made available in Rust)
+pub trait BatchValues: for<'r> BatchValuesGatWorkaround<'r> {}
+impl<T: for<'r> BatchValuesGatWorkaround<'r> + ?Sized> BatchValues for T {}
+
+pub trait BatchValuesGatWorkaround<'r, ImplicitBounds = &'r Self> {
+    type BatchValuesIter: BatchValuesIterator<'r>;
+    fn batch_values_iter(&'r self) -> Self::BatchValuesIter;
+    fn len(&'r self) -> usize;
```

Comment:
It is used there: https://github.com/scylladb/scylla-rust-driver/blob/a818ae55c3c08d12083c67ebae66d4ab425c0462/scylla/src/transport/connection.rs#L645

Reply:
OK. I think this check can be removed. It is redundant because you added a later check that compares statement and value counts match during serialization (in `Batch::serialize`). This error should only happen if the API is misused, so I think it is fine to detect it later, even at a point when we have already spent some time serializing the request.

Reply:
IIRC there also the type that differs at that point so we can't return the user-friendly error without more significant changes. Since the user has to know the count in values so that it matches the count in statements I feel like keeping the len requirement is unlikely to be a significant constraint.
It should also be possible to update the serialization error so that it may propagate that error back in a typed manner, though I'd have to look into it a bit to be sure of the extent of the interface update that this would require.

Reply:
Updated the interface this way: bc80d1a3503ba1cc90d15063f39e112c856c51d9